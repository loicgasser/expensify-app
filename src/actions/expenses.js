import database from '../firebase/firebase'
import { createDeflate } from 'zlib';

// Add expense

export const addExpense = (expense) => (
    {
        type: 'ADD_EXPENSE',
        expense
    }
)

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

// Remove expense

export const removeExpense = ({ id } = {}) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
)

// Edit Expense

export const editExpense = (id, updates) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates

    }
)

// Set Expenses

export const setExpenses = (expenses) => (
    {
        type: 'SET_EXPENSES',
        expenses
    }
)

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((dataSnapshot) => {
            const expenses = []
            const expensesData = dataSnapshot.val()
            Object.keys(expensesData).forEach((key) => {
                const { description, amount, createdAt, note } = expensesData[key]
                expenses.push({
                    id: key,
                    description,
                    amount,
                    createdAt,
                    note
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}