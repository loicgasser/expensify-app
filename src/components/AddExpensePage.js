import React from 'react'

const AddExpensePage = () => (
    <div>This is my expense component.</div>
)

export default AddExpensePage