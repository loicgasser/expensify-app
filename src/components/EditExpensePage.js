import React from 'react'

const EditExpensePage = (props) => (
    <div>This is my edit component. My item id is {props.id}</div>
)

export default EditExpensePage