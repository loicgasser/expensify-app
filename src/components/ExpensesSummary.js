import React from 'react'
import { connect } from 'react-redux'
import sumExpenses from '../selector/expenses-total'
import numeral from 'numeral'

export class ExpensesSummary extends React.Component {
    render() {
        return (
            <p>
                Viewing {this.props.countExpenses} expenses totalling {numeral(this.props.sumExpenses / 100).format('$0,0.00')}
            </p>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        sumExpenses: sumExpenses(props.expenses),
        countExpenses: props.expenses.length
    }
}


export default connect(mapStateToProps)(ExpensesSummary)