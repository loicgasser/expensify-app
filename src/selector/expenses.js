// Function to filter and sort expenses getVisibleExpenses
import moment from 'moment'

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        const createdAt = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt, 'day') : true
        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            // Order most recent to oldest
            return a.createdAt > b.createdAt ? -1 : 1
        } else if (sortBy == 'amount') {
            // Reverse order, biggest amount first
            return a.amount > b.amount ? -1 : 1
        }
    })
}