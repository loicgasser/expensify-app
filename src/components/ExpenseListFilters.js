import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {
    state = {
        focusedInput: null
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        const value = e.target.value
        if (value === "date") {
            this.props.sortByDate()
        } else if (value === "amount") {
            this.props.sortByAmount()
        }
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (focusedInput) => {
        this.setState({ focusedInput })
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            placeholder="Search expenses"
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            values={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId="startDateId"
                            endDate={this.props.filters.endDate}
                            endDateId="endDateId"
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.focusedInput}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)