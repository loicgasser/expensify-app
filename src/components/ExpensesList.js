import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import ExpenseListFilters from './ExpenseListFilters'
import getVisibleExpenses from '../selector/expenses'

export const ExpensesList = (props) => (
    <div>
        <ExpenseListFilters />
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem
                        key={expense.id}
                        {...expense}
                    />
                })
            )
        }
        
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpensesList)