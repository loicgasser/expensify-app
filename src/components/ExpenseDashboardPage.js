import React from 'react'
import ExpensesList from './ExpensesList'

const ExpenseDashboardPage = () => (
    <div>
        <p>This is my dashboard component.</p>
        <ExpensesList />
    </div>
)

export default ExpenseDashboardPage