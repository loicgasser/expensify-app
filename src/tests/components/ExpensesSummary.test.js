import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})