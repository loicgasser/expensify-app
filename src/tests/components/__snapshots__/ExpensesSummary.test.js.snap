// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with expenses 1`] = `
<p>
  Viewing 
   expenses totalling 
  $0.00
</p>
`;

exports[`should render ExpensesSummary with no expenses 1`] = `
<p>
  Viewing 
   expenses totalling 
  $0.00
</p>
`;
